be2-scala-dup/src/main/scala/ch/epfl/pop/decentralized/GossipManager.scala

```diff
@@ -0,0 +1,187 @@
+package ch.epfl.pop.decentralized
+
+import akka.NotUsed
+import akka.actor.{Actor, ActorLogging, ActorRef, Props}
+import akka.event.slf4j.Logger
+import akka.pattern.AskableActorRef
+import akka.stream.scaladsl.Flow
+import ch.epfl.pop.decentralized.{ConnectionMediator, GossipManager}
+import ch.epfl.pop.model.network.method.message.Message
+import ch.epfl.pop.model.network.method.{GreetServer, Rumor}
+import ch.epfl.pop.model.network.{JsonRpcRequest, JsonRpcResponse, MethodType}
+import ch.epfl.pop.model.objects.{Base64Data, Channel, PublicKey}
+import ch.epfl.pop.pubsub.AskPatternConstants
+import ch.epfl.pop.pubsub.ClientActor.ClientAnswer
+import ch.epfl.pop.pubsub.graph.validators.RpcValidator
+import ch.epfl.pop.pubsub.graph.{ErrorCodes, GraphMessage, PipelineError}
+import ch.epfl.pop.storage.DbActor
+import ch.epfl.pop.storage.DbActor.DbActorReadRumor
+
+import scala.concurrent.Await
+import scala.util.{Random, Success}
+
+final case class GossipManager(
+    dbActorRef: AskableActorRef,
+    monitorRef: ActorRef,
+    connectionMediator: AskableActorRef,
+    stopProbability: Double = 0.5
+) extends Actor with AskPatternConstants with ActorLogging {
+
+  private type ServerInfos = (ActorRef, GreetServer)
+  private val publicKey: Option[PublicKey] = {
+    val readPk = dbActorRef ? DbActor.ReadServerPublicKey()
+    Await.result(readPk, duration) match
+      case DbActor.DbActorReadServerPublicKeyAck(pk) => Some(pk)
+      case _                                         => None
+  }
+  private var rumorId = 0
+  private var jsonId = 0
+
+  monitorRef ! GossipManager.Ping()
+  connectionMediator ? GossipManager.Ping()
+
+  private var activeGossipProtocol: Map[JsonRpcRequest, List[ServerInfos]] = Map.empty
+
+  private def isRumorNew(rumor: Rumor): Boolean = {
+    val readRumorDb = dbActorRef ? DbActor.ReadRumor(rumor.senderPk -> rumor.rumorId)
+    Await.result(readRumorDb, duration) match
+      case DbActorReadRumor(foundRumors) =>
+        foundRumors.isEmpty
```

Comment:
the match is not exhaustive here.

Reply:
I am removing this method as it is not necessary anymore